File: .fullsend/skills/e2e-failure-analysis/scripts/diagnostics.py
#!/usr/bin/env python3
"""Extract test results, deployment warnings, and config dumps from results.json.

Usage:
    python3 diagnostics.py <ARTIFACTS_DIR> [--project <name>]

Arguments:
    ARTIFACTS_DIR   Path to the artifacts directory containing playwright-report/results.json
    --project       Filter output to a specific Playwright project name (case-insensitive substring match)

Output sections:
    1. TEST RESULTS — pass/fail/skip counts and failed test details
    2. DIAGNOSTICS PER FAILED PROJECT — config table dumps and filtered warnings/errors
"""

import json
import re
import sys
from pathlib import Path


def walk_suites(suite, specs=None):
    if specs is None:
        specs = []
    for spec in suite.get("specs", []):
        specs.append(spec)
    for child in suite.get("suites", []):
        walk_suites(child, specs)
    return specs


WARNING_PATTERNS = re.compile(
    r"YAML file.*does not exist|"
    r"error|Error|ERROR|"
    r"WARN|warn|Warning|"
    r"does not exist|not found|"
    r"missing|Missing|"
    r"failed|Failed|FAILED|"
    r"CrashLoopBackOff|ImagePullBackOff|"
    r"secret.*not|Secret.*not",
    re.IGNORECASE,
)
NOISE = re.compile(r"node_modules|trace_id|Download|Extracting|integrity checksum")


def main():
    if len(sys.argv) < 2:
        print(f"Usage: {sys.argv[0]} <ARTIFACTS_DIR> [--project <name>]", file=sys.stderr)
        sys.exit(1)

    artifacts = Path(sys.argv[1])
    project_filter = None
    if "--project" in sys.argv:
        idx = sys.argv.index("--project")
        if idx + 1 < len(sys.argv):
            project_filter = sys.argv[idx + 1].lower()

    results_file = artifacts / "playwright-report" / "results.json"
    if not results_file.exists():
        print(f"ERROR: {results_file} not found", file=sys.stderr)
        sys.exit(1)

    r = json.loads(results_file.read_text())
    stats = r.get("stats", {})
    all_specs = []
    for s in r.get("suites", []):
        walk_suites(s, all_specs)

    # Section 1: Test Results
    print("=" * 70)
    print("TEST RESULTS")
    print("=" * 70)
    print(
        f"Passed: {stats.get('expected', 0)}  "
        f"Failed: {stats.get('unexpected', 0)}  "
        f"Skipped: {stats.get('skipped', 0)}  "
        f"Flaky: {stats.get('flaky', 0)}"
    )
    print()

    failed_projects = set()
    for spec in all_specs:
        for test in spec.get("tests", []):
            proj = test.get("projectName", "?")
            if project_filter and project_filter not in proj.lower():
                continue
            for result in test.get("results", []):
                status = result["status"]
                if status not in ("passed", "skipped"):
                    failed_projects.add(proj)
                    err = result.get("error", {}).get("message", "no error")
                    print(f"FAILED [{proj}] {spec['title']}")
                    print(f"  {err[:400]}")
                    print()
                elif project_filter:
                    print(f"{status.upper()} [{proj}] {spec['title']}")

    # Section 2: Diagnostics per failed project
    for proj in sorted(failed_projects):
        print("=" * 70)
        print(f"DIAGNOSTICS FOR PROJECT: {proj}")
        print("=" * 70)
        for spec in all_specs:
            for test in spec.get("tests", []):
                if test.get("projectName") != proj:
                    continue
                for result in test.get("results", []):
                    stdout = "".join(
                        s.get("text", "") for s in result.get("stdout", [])
                    )
                    if not stdout:
                        continue

                    title = spec["title"][:50]

                    # Config tables
                    for match in re.finditer(
                        r"(┌─[^\n]*(?:Config|Plugins|Deployment|Value|Backstage)[^\n]*\n)(.*?)(└─+)",
                        stdout,
                        re.DOTALL,
                    ):
                        header = match.group(1).strip()
                        body = match.group(2)
                        if len(body) < 5000:
                            print(f"--- {header} [{title}] ---")
                            print(body.rstrip())
                            print()

                    # Warnings and errors
                    warnings = []
                    for line in stdout.split("\n"):
                        if WARNING_PATTERNS.search(line) and not NOISE.search(line):
                            warnings.append(line.rstrip())
                    if warnings:
                        print(f"--- Warnings/Errors [{title}] ---")
                        for w in warnings[-40:]:
                            print(w)
                        print()


if __name__ == "__main__":
    main()

File: .fullsend/skills/e2e-failure-analysis/scripts/diagnostics.ts
#!/usr/bin/env node
import * as fs from "node:fs";
import * as path from "node:path";

interface Spec {
  title: string;
  tests?: Test[];
}

interface Test {
  projectName?: string;
  results?: TestResult[];
}

interface TestResult {
  status: string;
  error?: { message?: string };
  stdout?: Array<{ text?: string }>;
}

interface Suite {
  specs?: Spec[];
  suites?: Suite[];
}

interface ResultsJSON {
  stats?: {
    expected?: number;
    unexpected?: number;
    skipped?: number;
    flaky?: number;
  };
  suites?: Suite[];
}

function walkSuites(suite: Suite, specs: Spec[] = []): Spec[] {
  for (const spec of suite.specs || []) {
    specs.push(spec);
  }
  for (const child of suite.suites || []) {
    walkSuites(child, specs);
  }
  return specs;
}

const WARNING_PATTERNS = new RegExp(
  "YAML file.*does not exist|" +
  "error|Error|ERROR|" +
  "WARN|warn|Warning|" +
  "does not exist|not found|" +
  "missing|Missing|" +
  "failed|Failed|FAILED|" +
  "CrashLoopBackOff|ImagePullBackOff|" +
  "secret.*not|Secret.*not",
  "i",
);
const NOISE = /node_modules|trace_id|Download|Extracting|integrity checksum/;

function main(): void {
  if (process.argv.length < 3) {
    process.stderr.write(`Usage: ${process.argv[1]} <ARTIFACTS_DIR> [--project <name>]\n`);
    process.exit(1);
  }

  const artifacts = process.argv[2];
  let projectFilter: string | null = null;
  const projIdx = process.argv.indexOf("--project");
  if (projIdx !== -1 && projIdx + 1 < process.argv.length) {
    projectFilter = process.argv[projIdx + 1].toLowerCase();
  }

  const resultsFile = path.join(artifacts, "playwright-report", "results.json");
  if (!fs.existsSync(resultsFile)) {
    process.stderr.write(`ERROR: ${resultsFile} not found\n`);
    process.exit(1);
  }

  const r: ResultsJSON = JSON.parse(fs.readFileSync(resultsFile, "utf-8"));
  const stats = r.stats || {};
  const allSpecs: Spec[] = [];
  for (const s of r.suites || []) {
    walkSuites(s, allSpecs);
  }

  // Section 1: Test Results
  console.log("=".repeat(70));
  console.log("TEST RESULTS");
  console.log("=".repeat(70));
  console.log(
    `Passed: ${stats.expected || 0}  ` +
    `Failed: ${stats.unexpected || 0}  ` +
    `Skipped: ${stats.skipped || 0}  ` +
    `Flaky: ${stats.flaky || 0}`,
  );
  console.log();

  const failedProjects = new Set<string>();
  for (const spec of allSpecs) {
    for (const test of spec.tests || []) {
      const proj = test.projectName || "?";
      if (projectFilter && !proj.toLowerCase().includes(projectFilter)) {
        continue;
      }
      for (const result of test.results || []) {
        const status = result.status;
        if (status !== "passed" && status !== "skipped") {
          failedProjects.add(proj);
          const err = result.error?.message || "no error";
          console.log(`FAILED [${proj}] ${spec.title}`);
          console.log(`  ${err.slice(0, 400)}`);
          console.log();
        } else if (projectFilter) {
          console.log(`${status.toUpperCase()} [${proj}] ${spec.title}`);
        }
      }
    }
  }

  // Section 2: Diagnostics per failed project
  for (const proj of [...failedProjects].sort()) {
    console.log("=".repeat(70));
    console.log(`DIAGNOSTICS FOR PROJECT: ${proj}`);
    console.log("=".repeat(70));
    for (const spec of allSpecs) {
      for (const test of spec.tests || []) {
        if (test.projectName !== proj) continue;
        for (const result of test.results || []) {
          const stdout = (result.stdout || []).map((s) => s.text || "").join("");
          if (!stdout) continue;

          const title = spec.title.slice(0, 50);

          // Config tables
          const tableRe =
            /(┌─[^\n]*(?:Config|Plugins|Deployment|Value|Backstage)[^\n]*\n)(.*?)(└─+)/gs;
          let match;
          while ((match = tableRe.exec(stdout)) !== null) {
            const header = match[1].trim();
            const body = match[2];
            if (body.length < 5000) {
              console.log(`--- ${header} [${title}] ---`);
              console.log(body.trimEnd());
              console.log();
            }
          }

          // Warnings and errors
          const warnings: string[] = [];
          for (const line of stdout.split("\n")) {
            if (WARNING_PATTERNS.test(line) && !NOISE.test(line)) {
              warnings.push(line.trimEnd());
            }
          }
          if (warnings.length > 0) {
            console.log(`--- Warnings/Errors [${title}] ---`);
            for (const w of warnings.slice(-40)) {
              console.log(w);
            }
            console.log();
          }
        }
      }
    }
  }
}

main();
